### PR TITLE
WIP for review.

### PR DESCRIPTION
**Pull request by owais-vd**
Hi @bennsimon I have completed the for review.

### CODE REVIEW

File: opensrp-pnc/src/main/java/org/smartregister/pnc/PncLibrary.java [hunk 1]
```diff
@@ -238,4 +260,11 @@ public PncRulesEngineHelper getPncRulesEngineHelper() {
     protected Date getDateNow() {
         return new Date();
     }
+
+    public AppExecutors getAppExecutors() {
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/PncLibrary.java:256]: If this is here, kindly replace all the other uses of `AppExecutors` in the app to use it, instead of local instantiation in classes.

File: opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncStillBornRepository.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.smartregister.pnc.dao.PncGenericDao;
+import org.smartregister.pnc.pojo.PncStillBorn;
+import org.smartregister.pnc.utils.PncDbConstants;
+import org.smartregister.repository.BaseRepository;
+
+import java.util.List;
+
+public class PncStillBornRepository extends BaseRepository implements PncGenericDao<PncStillBorn> {
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncStillBornRepository.java]: This is not needed, in maternity its combined to one table i.e Maternity Births (Maternity live births + Maternity still born)

File: opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncUtils.java [hunk 3]
```diff
@@ -298,4 +303,77 @@ public static JSONArray generateFieldsFromJsonForm(@NonNull JSONObject jsonFormO
         return jsonArray;
     }
 
+    public static HashMap<String, HashMap<String, String>> buildRepeatingGroupTests(@NonNull JSONObject stepJsonObject, String fieldName) throws JSONException {
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncUtils.java]: kindly update the name of this method

File: opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncUtils.java [hunk 4]
```diff
@@ -298,4 +303,77 @@ public static JSONArray generateFieldsFromJsonForm(@NonNull JSONObject jsonFormO
...
+                    }
+                }
+            }
+        }
+        return repeatingGroupMap;
+    }
+
+    public static HashMap<String, String> getPncClient(String baseEntityId) {
+        ArrayList<HashMap<String, String>> hashMap;
+        hashMap = CoreLibrary.getInstance().context().getEventClientRepository().rawQuery(PncLibrary.getInstance().getRepository().getReadableDatabase(),
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncUtils.java]: you can combine declaration and assigning

File: sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java [hunk 5]
```diff
@@ -33,7 +33,7 @@ public String getObjectIdsQuery(@Nullable String filters, @Nullable String mainC
     public String[] countExecuteQueries(@Nullable String filters, @Nullable String mainCondition) {
         SmartRegisterQueryBuilder sqb = new SmartRegisterQueryBuilder();
         return new String[] {
-                sqb.countQueryFts("ec_client", null, "register_type = 'pnc'", filters)
+                sqb.countQueryFts("ec_client", null, null, filters)
```
Comment (bennsimon) [inline sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java]: update this to use table specified in `PncMetadata`

File: sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java [hunk 6]
```diff
@@ -18,10 +18,10 @@
     @Override
     public String getObjectIdsQuery(@Nullable String filters, @Nullable String mainCondition) {
         if (TextUtils.isEmpty(filters)) {
-            return "SELECT object_id, last_interacted_with FROM ec_client_search WHERE register_type = 'pnc' " +
+            return "SELECT object_id, last_interacted_with FROM ec_client_search " +
                     "ORDER BY last_interacted_with DESC";
         } else {
-            String sql = "SELECT object_id FROM ec_client_search WHERE date_removed IS NULL AND register_type = 'pnc' AND phrase MATCH '%s*'" +
+            String sql = "SELECT object_id FROM ec_client_search WHERE date_removed IS NULL AND phrase MATCH '%s*'" +
                     "ORDER BY last_interacted_with DESC";
```
Comment (bennsimon) [inline sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java]: update the table to table specified in `PncMetadata` and for fts table occurrences e.g `ec_client_search`  use `CommonFtsObject.searchTableName(table from Pnc Metadata )`

**Review by bennsimon — CHANGES_REQUESTED** [review]

**Comment by bennsimon** [comment]
Kindly remove this line https://github.com/OpenSRP/opensrp-client-pnc/blob/3c2f96f5944e7b4a6d3a4db067a3ee7a52438d35/settings.gradle#L1

**Comment by bennsimon** [comment]
Use repeating groups on step 3 to feed in children born and/or stillbirths

File: opensrp-pnc/src/main/java/org/smartregister/pnc/PncLibrary.java (hunk 1 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/PncLibrary.java:256]: refactor all AppExecutors with this

File: opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncStillBornRepository.java (hunk 2 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncStillBornRepository.java]: deleted.

File: opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncUtils.java (hunk 3 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncUtils.java]: removed test postfix

File: opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncUtils.java (hunk 4 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncUtils.java]: done

File: sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java (hunk 5 above)
Comment (owais-vd) [reply sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java]: updated.

File: sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java (hunk 6 above)
Comment (owais-vd) [reply sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java]: done

File: opensrp-pnc/src/main/java/org/smartregister/pnc/PncLibrary.java (hunk 1 above)
Comment (bennsimon) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/PncLibrary.java:256]: yeah `PncLibrary.getInstance().getAppExecutors()`

File: sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java (hunk 6 above)
Comment (bennsimon) [reply sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java]: you can use this `PncUtils.metadata().getTableName()`

File: opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncDbConstants.java [hunk 7]
```diff
@@ -54,12 +53,39 @@
...
+        }
     }
 
     interface Table {
         String EC_CLIENT = "ec_client";
         String PNC_DETAILS = "pnc_details";
         String PNC_REGISTRATION_DETAILS = "pnc_registration_details";
         String PNC_OUTCOME_FORM = "pnc_outcome_form";
+        String PNC_BABY = "pnc_baby";
+        String PNC_STILL_BORN = "pnc_still_born";
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncDbConstants.java]: you can remove this

File: opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncDbConstants.java [hunk 8]
```diff
@@ -54,12 +53,39 @@
...
+            String BF_FIRST_HOUR = "bf_first_hour";
+            String INTERVENTION_REFERRAL_LOCATION = "intervention_referral_location";
+            String INTERVENTION_SPECIFY = "intervention_specify";
+            String CARE_MGT = "care_mgt";
+        }
+
+        interface PncStillBorn {
+            String MOTHER_BASE_ENTITY_ID = "mother_base_entity_id";
+            String STILLBIRTH__CONDITION = "stillbirth_condition";
+        }
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncDbConstants.java]: remove this

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java (hunk 6 above)
Comment (owais-vd) [reply sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java]: Done

File: opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncDbConstants.java (hunk 7 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncDbConstants.java]: PNC_STILL_BORN removed. Whereas PNC_BABY used in ChildRepository

File: opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncDbConstants.java (hunk 8 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/utils/PncDbConstants.java]: removed

File: opensrp-pnc/src/main/assets/json.form/pnc_registration.json [hunk 9]
```diff
@@ -50,1173 +50,2596 @@
...
+          "label": "Marital status",
+          "label_text_style": "bold",
+          "options": [
+            {
+              "key": "married_living_together",
+              "text": "Married & living together",
+              "openmrs_entity_parent": "",
+              "openmrs_entity": "concept",
+              "openmrs_entity_id": "1055AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+            },
```
Comment (bennsimon) [inline opensrp-pnc/src/main/assets/json.form/pnc_registration.json]: update this from the data dictionary

File: opensrp-pnc/src/main/java/org/smartregister/pnc/activity/BasePncFormActivity.java [hunk 10]
```diff
@@ -77,7 +77,7 @@ public void initializeFormFragment() {
 
     protected void initializeFormFragmentCore() {
         BasePncFormFragment pncFormFragment = (BasePncFormFragment) BasePncFormFragment.getFormFragment(JsonFormConstants.FIRST_STEP_NAME);
-        getSupportFragmentManager().beginTransaction().add(com.vijay.jsonwizard.R.id.container, pncFormFragment).commit();
+        getSupportFragmentManager().beginTransaction().add(com.vijay.jsonwizard.R.id.container, pncFormFragment).addToBackStack(pncFormFragment.getArguments().getString(JsonFormConstants.STEPNAME)).commit();
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/activity/BasePncFormActivity.java]: you can revert this change ... its no longer needed

File: opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java [hunk 11]
```diff
@@ -48,12 +67,185 @@ public boolean canProcess(@NonNull String eventType) {
...
+
+            HashMap<String, String> keyValues = new HashMap<>();
+            generateKeyValuesFromEvent(event, keyValues, true);
+
+            PncRegistrationDetails maternityDetails = new PncRegistrationDetails(eventClient.getClient().getBaseEntityId(), event.getEventDate().toDate(), keyValues);
+            maternityDetails.setCreatedAt(new Date());
+
+            PncLibrary.getInstance().getPncRegistrationDetailsRepository().saveOrUpdate(maternityDetails);
+//            processEvent(eventClient.getEvent(), eventClient.getClient(), clientClassification);
+//            CoreLibrary.getInstance().context().getEventClientRepository().markEventAsProcessed(eventClient.getEvent().getFormSubmissionId());
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java:109]: Could you update this logic to be as that of maternity .... just on the registration and update registration events bit. Then clean up any additional code

File: opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java [hunk 12]
```diff
@@ -48,12 +67,185 @@ public boolean canProcess(@NonNull String eventType) {
...
+            }
+            processEvent(eventClient.getEvent(), eventClient.getClient(), clientClassification);
+            CoreLibrary.getInstance().context().getEventClientRepository().markEventAsProcessed(eventClient.getEvent().getFormSubmissionId());
+            unsyncEvents.add(event);
+        } else if (eventType.equals(PncConstants.EventTypeConstants.PNC_OUTCOME)) {
+            processEvent(eventClient.getEvent(), eventClient.getClient(), clientClassification);
+            processPncOutcome(event);
+            CoreLibrary.getInstance().context().getEventClientRepository().markEventAsProcessed(eventClient.getEvent().getFormSubmissionId());
+        }*/
+
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java:87]: you can remove these commented out code

File: opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java [hunk 13]
```diff
@@ -48,12 +67,185 @@ public boolean canProcess(@NonNull String eventType) {
...
+                    pncChild.setFirstName(jsonTestObject.optString("baby_first_name"));
+                    pncChild.setLastName(jsonTestObject.optString("baby_last_name"));
+                    pncChild.setGender(jsonTestObject.optString("baby_gender"));
+                    pncChild.setChildHivStatus(jsonTestObject.optString("child_hiv_status"));
+                    pncChild.setHeight(jsonTestObject.optString("birth_height_entered"));
+                    pncChild.setWeight(jsonTestObject.optString("birth_weight_entered"));
+                    pncChild.setEventDate(PncUtils.convertDate(event.getEventDate().toDate(), PncDbConstants.DATE_FORMAT));
+                    pncChild.setNvpAdministration(jsonTestObject.optString("nvp_administration"));
+                    pncChild.setInterventionSpecify(jsonTestObject.optString("baby_intervention_specify"));
+                    pncChild.setInterventionReferralLocation(jsonTestObject.optString("baby_intervention_referral_location"));
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java:164]: add these strings to constants and use its references here

File: opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java [hunk 14]
```diff
@@ -48,12 +67,185 @@ public boolean canProcess(@NonNull String eventType) {
...
+    private void processStillBorn(@Nullable String strStillBorn, @NonNull Event event) {
+        if (StringUtils.isNotBlank(strStillBorn)) {
+            try {
+                JSONObject jsonObject = new JSONObject(strStillBorn);
+                Iterator<String> repeatingGroupKeys = jsonObject.keys();
+                while (repeatingGroupKeys.hasNext()) {
+                    JSONObject jsonTestObject = jsonObject.optJSONObject(repeatingGroupKeys.next());
+                    PncChild maternityStillBorn = new PncChild();
+                    maternityStillBorn.setMotherBaseEntityId(event.getBaseEntityId());
+                    maternityStillBorn.setStillBirthCondition(jsonTestObject.optString("stillbirth_condition"));
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java:182]: add this to a constants file

File: opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncRegistrationDetailsRepository.java [hunk 15]
```diff
@@ -14,10 +14,12 @@
 
 public class PncRegistrationDetailsRepository extends PncDetailsRepository {
 
+    private static final String TABLE = PncDbConstants.Table.PNC_REGISTRATION_DETAILS;
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncRegistrationDetailsRepository.java:17]: This file is not necessary

File: sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java [hunk 16]
```diff
@@ -33,7 +36,7 @@ public String getObjectIdsQuery(@Nullable String filters, @Nullable String mainC
     public String[] countExecuteQueries(@Nullable String filters, @Nullable String mainCondition) {
         SmartRegisterQueryBuilder sqb = new SmartRegisterQueryBuilder();
         return new String[] {
-                sqb.countQueryFts("ec_client", null, "register_type = 'pnc'", filters)
+                sqb.countQueryFts(PncLibrary.getInstance().getPncConfiguration().getPncMetadata().getTableName(), null, null, filters)
```
Comment (bennsimon) [inline sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java]: update this to `PncUtils.metadata().getTableName()`

File: opensrp-pnc/src/main/java/org/smartregister/pnc/pojo/PncStillBorn.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
+package org.smartregister.pnc.pojo;
+
+public class PncStillBorn {
```
Comment (bennsimon) [inline opensrp-pnc/src/main/java/org/smartregister/pnc/pojo/PncStillBorn.java:3]: this file is not necessary

**Review by bennsimon — CHANGES_REQUESTED** [review]

**Comment by bennsimon** [comment]
Remove the 3rd step of PNC registration. You only need 2

File: opensrp-pnc/src/main/assets/json.form/pnc_registration.json (hunk 9 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/assets/json.form/pnc_registration.json]: updated.

File: opensrp-pnc/src/main/java/org/smartregister/pnc/activity/BasePncFormActivity.java (hunk 10 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/activity/BasePncFormActivity.java]: reverted.

File: opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java (hunk 11 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java:109]: I have updated the file according to the need in issue 11 branch

File: opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java (hunk 12 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java:87]: I have updated the file according to the need in issue 11 branch

File: opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java (hunk 13 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java:164]: I have updated the file according to the need in issue 11 branch

File: opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncRegistrationDetailsRepository.java (hunk 15 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncRegistrationDetailsRepository.java:17]: I have used this for pnc outcome details

File: sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java (hunk 16 above)
Comment (owais-vd) [reply sample-app/src/main/java/org/smartregister/pnc/sample/config/PncRegisterQueryProvider.java]: updated.

File: opensrp-pnc/src/main/java/org/smartregister/pnc/pojo/PncStillBorn.java (hunk 17 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/pojo/PncStillBorn.java:3]: i have used this file in still birth section

File: opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncRegistrationDetailsRepository.java (hunk 15 above)
Comment (bennsimon) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncRegistrationDetailsRepository.java:17]: you can also remove the `PncOutcomeDetailsRepository`, but i guess you can include this to to issue 11

File: opensrp-pnc/src/main/assets/json.form/pnc_registration.json [hunk 18]
```diff
@@ -50,1173 +50,2603 @@
...
+            "step2:religion": {
+              "type": "string",
+              "ex": "equalTo(., \"other\")"
+            }
+          }
+        }
+
+      ]
+    },
+    "step3": {
```
Comment (bennsimon) [inline opensrp-pnc/src/main/assets/json.form/pnc_registration.json:586]: remove this step

File: opensrp-pnc/src/main/assets/json.form/pnc_registration.json [hunk 19]
```diff
@@ -1,9 +1,9 @@
 {
-  "count": "4",
-  "encounter_type": "PNC Registration",
-  "entity_id": "",
-  "relational_id": "",
-  "metadata": {
+    "count": "3",
```
Comment (bennsimon) [inline opensrp-pnc/src/main/assets/json.form/pnc_registration.json:2]: update this count to 2

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java (hunk 14 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/processor/PncMiniClientProcessorForJava.java:182]: fixed in branch 11

File: opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncRegistrationDetailsRepository.java (hunk 15 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/java/org/smartregister/pnc/repository/PncRegistrationDetailsRepository.java:17]: just ignore it right now because i have worked on that file in issue 11 and if I remove this one then this branch will conflict with issue 11 branch.

File: opensrp-pnc/src/main/assets/json.form/pnc_registration.json (hunk 18 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/assets/json.form/pnc_registration.json:586]: it has been already removed in issue 11. just ignore it. you will not see this in next PR

File: opensrp-pnc/src/main/assets/json.form/pnc_registration.json (hunk 19 above)
Comment (owais-vd) [reply opensrp-pnc/src/main/assets/json.form/pnc_registration.json:2]: same as above comments

**Review by bennsimon — APPROVED** [review]
